Handlers can have required params with extension points.

feerrenrut opened the pull request:
I figured I would try to address this before there are too many more usages of extension points.

### Link to issue number:
None

### Summary of the issue:
Extension points handlers must provide default values for arguments. This forces the handlers to treat arguments as optional, which sometimes does not make any sense.

### Description of how this pull request fixes the issue:
- Introduce several more unit tests.
- Try to clarify intent of existing unit tests.
- Add logic to allow handlers to have "required parameters". These are
parameters with no default value set.
- Add logic to catch and raise an exception when:
  - The required parameters are not supplied (either through positional
  args or keyword args)
  - The handler does not accept enough positional args.

The main point of this PR is to allow handlers to declare parameters without default values.
Default values indicate that args are perhaps optional. Even if the handlers do provide default values, we could not expect to be able to remove any specific keyword args from an Action/Filter/Decider and maintain backwards compatibility.

This PR also adds a few more tests around various combinations of handlers with default values / accepting kwargs, and calling callWithSupportedKwargs with positional arguments. I added checks for, and exceptions for some situations where parameters do not match up well.

I'm a bit iffy on having callWithSupportedKwargs being called with positional arguments. It would be easy for the order to be messed up. This may happen in a case where a new location in the code is added to trigger a pre-existing action. Positional args do not provide a very well documented api for those registering handlers, when keywords are given it is more apparent what the values are supposed to represent. I can only imagine the reason to do this is convenience for the extension method implementer, while this saves some typed characters, I think it could easily come at the cost of clarity and time spent debugging.

If the performance of this code is a concern, I would suggest we move the argument matching to registration time. Though in that case, I would recommend we are more strict about handlers matching some signature. Perhaps a signature must be registered when the the extension point is created?

### Testing performed:
Unit tests.

### Known issues with pull request:
None

### Change log entry:
Not sure.

feerrenrut committed:
Handlers can have required params with extension points.

- Introduce several more unit tests.
- Try to clarify intent of existing unit tests.
- Add logic to allow handlers to have "required parameters". These are
parameters with no default value set.
- Add logic to catch and raise an exception when:
  - The required parameters are not supplied (either through positional
  args or keyword args)
  - The handler does not accept enough positional args.

feerrenrut commented:
@jcsteh I would be interested to hear your thoughts on this change. It's a lot to take in, so feel free to say you don't have time for this one.

jcsteh commented on source/extensionPoints.py:
```diff
@@ -119,16 +124,48 @@ def myFunc(a=None, b=None):
 	"""
 	spec = inspect.getargspec(func)
+
+	# Ensure that we can provide all arguments without defaults expected by the handler are provided.
```
This sentence needs some tweaking. :) Maybe remove the "can provide"?

jcsteh commented on source/extensionPoints.py:
```diff
@@ -119,16 +124,48 @@ def myFunc(a=None, b=None):
+	# if this tuple has n elements, they correspond to the last n elements listed in args.
+	numExpectedArgsWithDefaults = len(spec.defaults) if spec.defaults else 0
+
+	# some handlers are member functions, discard "self" because it is passed implicitly.
```
It might be better to use the term "instance method" rather than "member function" here, just because I see that more in the Python world. Also, this won't work for class methods, where the first argument is normally named "cls". I don't know if that's a scenario we need to care about, though.

jcsteh commented on source/extensionPoints.py:
```diff
@@ -119,16 +124,48 @@ def myFunc(a=None, b=None):
+	numExpectedArgs = len(spec.args)
+	numGivenPositionalArgs = len(args)
+	if numGivenPositionalArgs > numExpectedArgs:
+		raise Exception("Expected to be able to pass {} positional arguments.".format(numGivenPositionalArgs))
```
I'd be inclined to use TypeError instead of just Exception.

jcsteh commented on source/extensionPoints.py:
```diff
@@ -142,14 +179,17 @@ class Action(HandlerRegistrar):
 
-	Interested parties then register to be notified about this action:
+	Interested parties then register to be notified about this action, see
+	c@{HandleRegistrar.register} docstring for details of the type of handlers that can be
```
Note that in Epydoc syntax, you probably want to use `L{some.method}`, which will generate a link.

jcsteh commented on tests/unit/test_extensionPoints.py:
```diff
@@ -48,56 +48,259 @@ def test_isWeak(self):
 
 class TestCallWithSupportedKwargs(unittest.TestCase):
+	"""
+	Tests to ensure that the correct parameters/kwargs are passed through to the handler method.
```
I'm a bit worried about the term `kwargs` throughout. It's not clear to me when you mean capturing extra kwargs (`**kwargs`) vs named arguments (`a=1`). Perhaps you could use `**kwargs` to be explicit... or the term `extra kwargs` maybe?

jcsteh commented on tests/unit/test_extensionPoints.py:
```diff
@@ -48,56 +48,259 @@ def test_isWeak(self):
-			gotKwargs["b"] = b
-		extensionPoints.callWithSupportedKwargs(handler, a=1)
-		self.assertEqual(gotKwargs, {"a": 1, "b": 2})
+	def test_handlerTakesNoParams_someKwargsGiven_handlerIsCalled(self):
```
`someKwargsGiven` seems to break the pattern here (and a few other places below); previously, you have the words `takes` and `given` first. I know "some kwargs given" is more linguistically correct than "given some kwargs", but I wonder whether consistency might win out here.

jcsteh commented on tests/unit/test_extensionPoints.py:
```diff
@@ -48,56 +48,259 @@ def test_isWeak(self):
+
+	def test_handlerTakesParamsWithDefaultAndKwargs_otherKwargsGiven_handlerGetsOtherKwargsAndDefaultValues(self):
+		"""Test that **kwargs is still passed in when params aren't provided.
+		Handler has default vales for params, and takes kwargs.
```
vales -> values

jcsteh commented on tests/unit/test_extensionPoints.py:
```diff
@@ -48,56 +48,259 @@ def test_isWeak(self):
+		""" Basic test for callWithSupportedKwargs function.
+		Handler expects a single parameter, with a default provided.
+		callWithSupportedKwargs is called with a single kwarg that matches.
+		Expectaion: handler called with parameter value equal to kwargs given to callWithSupportedKwargs.
```
Expectaion -> Expectation

jcsteh commented on tests/unit/test_extensionPoints.py:
```diff
@@ -204,15 +407,28 @@ def handler2():
 		self.action.notify()
 		self.assertEqual(called, [handler1, handler2])
 
-	def test_kwargs(self):
```
Did you intend to remove this test (and the later test_kwargs tests)? I guess this is all tested by the `callWithSupportedKwargs` tests, but I see you do have member function tests here and I wonder whether it might still be useful to make sure this works for sanity sake, given that this is where callers will hit us.

jcsteh reviewed (commented):
We discussed the thinking behind this on the phone, and in general, I follow. The below is by no means a full review; I mostly read through the unit tests (which look good) and skimmed the rest. So, I'm not marking this as "approved"; the new code in `callWithSupportedKwargs` may need some closer scrutiny.

Note that this all gets a lot nicer in Python 3, since you can define required keyword arguments without defaults:

```
>>> def func(p1, *, k1):
...  print(p1, k1)
...
>>> func("p1")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: func() missing 1 required keyword-only argument: 'k1'
>>> func("p1", "k1")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: func() takes 1 positional argument but 2 were given
>>> func("p1", k1="k1")
p1 k1
```

Once we transition to Python 3, we should re-evaluate this stuff.

feerrenrut committed:
Initial review actions for #8208

feerrenrut committed:
Merge remote-tracking branch 'origin/master' into extensionPointsNonDefaultArgsForHandlers

feerrenrut committed:
Move extensionPoints into a package

This is to make it clearer where to start for someone new to this API

feerrenrut committed:
Fix pydoc

feerrenrut commented:
Thanks for the review @jcsteh, I have addressed your comments. As discussed on the phone, I have also made extension points into a package, with the common classes (`Action`, `Filter`, `Decider`) in `__init__`, and the rest of the code in a `util.py` file. I have tested this move by ensuring the tests still pass, and by manually running NVDA.
